### PR TITLE
Assign roles to members joining via token gates

### PR DESCRIPTION
**Pull request by Devorein**
1. A unified api to update the roles attached with a token gate
2. Any new member will be assigned the roles attached to the token gates
3. Any existing member will only be assigned the new roles (that haven't already been assigned) attached to the token gates
4. Relation between `role` and `tokenGateToRole` to delete the record whenever the attached role gets deleted
5. Only admins will be able to perform the above operations (validated both on server and client side)

TODO:
1. Removing user role if the wallet is no longer valid
2. UI update

### CODE REVIEW

File: charmClient.ts [hunk 1]
```diff
@@ -15,13 +15,15 @@ import { OctoUtils } from 'components/common/BoardEditor/focalboard/src/octoUtil
 import { InviteLinkPopulated } from 'pages/api/invites/index';
 import { FiatCurrency, IPairQuote } from 'models/Currency';
 import type { FailedImportsError } from 'lib/notion/types';
+// TODO: Maybe move these types to another place so that we dont import from backend
```
[the remark this answers is not in the document]
Comment (mattcasey) [reply charmClient.ts:18]: The types we use here should be specifically related to the api that CharmClient is using, that's why i put some here before. I don't think they necessarily belong in lib either, most of what's in lib should not care about what our frontend api looks like IMO. 
We do not import any components in this folder, only interfaces from the pages/api folder.

File: charmClient.ts [hunk 2]
```diff
@@ -532,6 +534,10 @@ class CharmClient {
   getPageThreads (pageId: string): Promise<ThreadWithComments[]> {
     return http.GET(`/api/pages/${pageId}/threads`);
   }
+
+  updateTokenGateRoles (tokenGateId: string, spaceId: string, roleIds: string[]) {
+    return http.POST<TokenGateToRole[]>(`/api/token-gates/${tokenGateId}/roles`, { spaceId, roleIds });
+  }
 }
```
[the remark this answers is not in the document]
Comment (mattcasey) [reply charmClient.ts:541]: +1 try to group things. maybe we should even add namespaces... `charmClient.tokenGates.create()` I copied this pattern from Focalboard, but they only have maybe 10-20 paths. this client is getting huge! (but i still like that all api calls are in one place :)

File: charmClient.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattcasey) [reply charmClient.ts:18]: [edit] i saw below you're also advocating to move the code out of the middleware folder ;)

**Review by mattcasey — APPROVED** [review]

File: charmClient.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Devorein) [reply charmClient.ts:18]: Great point. I will remove all these types our API folder to a more suitable `lib` folder in a separate PR

File: components/settings/token-gates/TokenGateRolesSelect.tsx [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  onChange: (roleIds: string[]) => void
+  selectedRoleIds: string[]
+  onDelete: (roleId: string) => void
+}
+
+export default function TokenGateRolesSelect ({ onDelete, selectedRoleIds, onChange }: Props) {
+  const { roles } = useRoles();
+
+  const rolesRecord: Record<string, ListSpaceRolesResponse> = useMemo(() => roles ? roles?.reduce((obj, role) => (
+    { ...obj, [role.id]: role }
```
[the remark this answers is not in the document]
Comment (Devorein) [reply components/settings/token-gates/TokenGateRolesSelect.tsx:40]: I agree. Updated in latest commit.

File: components/settings/token-gates/hooks/useTokenGates.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        setTokenGatesWithRoles(data.reduce((acc, cur) => ({ ...acc, [cur.id]: cur }), {}));
+        const results = await Promise.all(data.map(tokenGate => humanizeAccessControlConditions({
+          myWalletAddress: account || '',
+          accessControlConditions: (tokenGate.conditions as any)?.accessControlConditions || []
+        }).then(description => {
+          return {
+            description,
+            tokenGateId: tokenGate.id
+          };
+        })));
```
[the remark this answers is not in the document]
Comment (Devorein) [reply components/settings/token-gates/hooks/useTokenGates.tsx:44]: Thank you 😄 Updated

File: charmClient.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Devorein) [reply charmClient.ts:541]: Agreed. I always end up adding any charmClient method to the very bottom 😛

File: components/settings/token-gates/hooks/useTokenGates.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+export function TokenGatesProvider ({ spaceId, children }:{spaceId: string, children: ReactNode}) {
+  const [tokenGatesWithRoles, setTokenGatesWithRoles] = useState<IContext['tokenGatesWithRoles']>({});
+  const { data, mutate } = useSWR(`tokenGates/${spaceId}`, () => charmClient.getTokenGates({ spaceId }));
+  const [tokenGateDescriptions, setTokenGateDescriptions] = useState<IContext['tokenGateDescriptions']>({});
+  const { account } = useWeb3React();
+
+  useEffect(() => {
+    async function main () {
+      if (data) {
+        setTokenGatesWithRoles(data.reduce((acc, cur) => ({ ...acc, [cur.id]: cur }), {}));
```
[the remark this answers is not in the document]
Comment (Devorein) [reply components/settings/token-gates/hooks/useTokenGates.tsx:35]: Thats a lot more readable. Thank you for the suggestion.

File: pages/api/token-gates/[id]/roles.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import nc from 'next-connect';
+
+const handler = nc<NextApiRequest, NextApiResponse>({ onError, onNoMatch });
+
+handler.use(requireUser)
+  .use(requireKeys(['roleIds', 'spaceId'], 'body'))
+  .use(requireSpaceMembership({ adminOnly: true }))
+  .post(updateTokenGateRoles);
+
+async function updateTokenGateRoles (req: NextApiRequest, res: NextApiResponse<TokenGateToRole[]>) {
```
[the remark this answers is not in the document]
Comment (Devorein) [reply pages/api/token-gates/[id]/roles.ts:17]: Good point. I contemplated on doing that. Maybe we should write down these sorts of best practices somewhere.

File: pages/api/token-gates/index.ts [hunk 7]
```diff
@@ -34,7 +34,12 @@ async function saveTokenGate (req: NextApiRequest, res: NextApiResponse) {
   res.status(200).json(result);
 }
 
-async function getTokenGates (req: NextApiRequest, res: NextApiResponse<TokenGate[] | { error: string }>) {
+export interface TokenGateWithRoles extends TokenGate {
+  space: Space;
+  tokenGateToRoles: TokenGateToRole[];
+}
+
```
[the remark this answers is not in the document]
Comment (Devorein) [reply pages/api/token-gates/index.ts:41]: Will make a separate PR to fix importing types from API handlers

File: pages/api/token-gates/[id]/verify.ts [hunk 8]
```diff
@@ -39,8 +39,13 @@ async function verifyWallet (req: NextApiRequest, res: NextApiResponse) {
           spaceId: tokenGate.spaceId
         }
       });
+      const tokenGateToRoles = await prisma.tokenGateToRole.findMany({
```
[the remark this answers is not in the document]
Comment (Devorein) [reply pages/api/token-gates/[id]/verify.ts:42]: Agree. This would help a lot in unit testing.

**Pull request merged**